pkg/pipeline/benchmark.go

```diff
@@ -0,0 +1,211 @@
+	version,
+	benchmarkJobURL string,
+	benchmarkJobDurationMins int) (*dagger.Container, error) {
+	if _, err := p.benchmark(ctx, cncfProject, config, version, benchmarkJobURL, benchmarkJobDurationMins); err != nil {
+		log.Printf("benchmark failed: %v", err)
+	}
+
+	if _, err := p.delete(ctx, cncfProject, config, benchmarkJobURL); err != nil {
+		return nil, err
+	}
+
+	return p.container, nil
+}
+
+// Terminal returns dagger interactive terminal configured with kubeconfig
+// for trouble shooting.
+func (p *Pipeline) Terminal(ctx context.Context) (*dagger.Container, error) {
+	return p.withKubeconfig().Terminal(), nil
+}
+
+// benchmark creates the CNCF project resource, applies the benchmark tests and
+// waits for the benchmark to complete.
+func (p *Pipeline) benchmark(ctx context.Context,
+	cncfProject,
+	config,
+	version,
+	benchmarkJobURL string,
+	benchmarkJobDurationMins int) (*dagger.Container, error) {
+	// Create CNCF project resources.
+	if _, err := p.deploy(ctx, cncfProject, config, version); err != nil {
+		return nil, err
+	}
+
+	// Create benchmark job resources.
+	if _, err := p.exec(ctx, cmd.Apply(benchmarkJobURL)); err != nil {
+		return nil, err
+	}
+
+	// Wait for pods to be ready.
+	if _, err := p.exec(ctx, cmd.WaitForReadyPods(benchmarkNamespace)); err != nil {
+		return nil, err
+	}
+
+	if _, err := p.echo(ctx, fmt.Sprintf("waiting %d minutes for benchmark to complete", benchmarkJobDurationMins)); err != nil {
+		return nil, err
+	}
+
+	time.Sleep(time.Duration(benchmarkJobDurationMins) * time.Minute)
+
+	if _, err := p.echo(ctx, "benchmark complete"); err != nil {
+		return nil, err
+	}
+
+	return p.container, nil
+}
+
+// delete cleans up the cluster at the end of the pipeline run.
+func (p *Pipeline) delete(ctx context.Context, cncfProject, config, benchmarkJobURL string) (*dagger.Container, error) {
+	// Delete benchmark job resources.
+	if _, err := p.exec(ctx, cmd.Delete(benchmarkJobURL)); err != nil {
+		log.Printf("failed to delete benchmark job: %v", err)
+	}
+
+	fileName, fileContents, err := p.getManifestFile(ctx, cncfProject, config, "")
+	if err != nil {
+		return nil, err
+	}
+
+	// Delete CNCF project resources.
+	if _, err := p.execWithNewFile(ctx, fileName, fileContents, cmd.Delete(fileName)); err != nil {
+		return nil, err
+	}
+
+	return p.container, nil
+}
+
+// deploy deploys the CNCF project by applying the flux manifest.
+func (p *Pipeline) deploy(ctx context.Context, cncfProject, config, version string) (*dagger.Container, error) {
+	fileName, fileContents, err := p.getManifestFile(ctx, cncfProject, config, version)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err = p.execWithNewFile(ctx, fileName, fileContents, cmd.Apply(fileName)); err != nil {
+		return nil, err
+	}
+
+	// Allow time for pods to be created. This is needed because there is a
+	// delay while flux reconciles the manifest. Without it the following
+	// kubectl wait command will fail.
+	time.Sleep(15 * time.Second)
```

Comment:
Is this sleep interval a random guess or there is something we are referring to, like any docs?

Reply:
This is basically the same as we have in the current pipeline.
https://github.com/cncf-tags/green-reviews-tooling/blob/87bbf10320f714ae88e74312dce5f4419239e5aa/.github/workflows/benchmark-pipeline.yaml#L104

It's not pretty but pipeline will fail without it and we spent quite some time debugging it.

Reply:
okay